_posts/2019-06-25-dynamic_linear_models_tfprobability/dynamic_linear_models_tfprobability.Rmd

```diff
@@ -0,0 +1,805 @@
+    aes(ymin = pred_mean - 2 * pred_sd, ymax = pred_mean + 2 * pred_sd),
+    alpha = 0.2,
+    fill = "cyan"
+  ) +
+  theme(axis.title = element_blank())
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "AirPassengers, 12-months-ahead forecast."}
+knitr::include_graphics("images/airp_forecast.png")
+```
+
+A call to [sts_decompose_by_component](https://rstudio.github.io/tfprobability/reference/sts_decompose_by_component.html) yields the -- centered -- components, a linear trend and a seasonal factor:
+
+```{r}
+component_dists <-
+  ts_train %>% sts_decompose_by_component(model = model, parameter_samples = param_samples)
+
+seasonal_effect_means <- component_dists[[1]] %>% tfd_mean()
+seasonal_effect_sds <- component_dists[[1]] %>% tfd_stddev()
+linear_effect_means <- component_dists[[2]] %>% tfd_mean()
+linear_effect_sds <- component_dists[[2]] %>% tfd_stddev()
+
+with(tf$Session() %as% sess, {
+  c(
+    seasonal_effect_means,
+    seasonal_effect_sds,
+    linear_effect_means,
+    linear_effect_sds
+  ) %<-% sess$run(
+    list(
+      seasonal_effect_means,
+      seasonal_effect_sds,
+      linear_effect_means,
+      linear_effect_sds
+    )
+  )
+})
+
+components_df <- forecast_df %>%
+  add_column(seasonal_effect_means = c(seasonal_effect_means, rep(NA, n_forecast_steps))) %>%
+  add_column(seasonal_effect_sds = c(seasonal_effect_sds, rep(NA, n_forecast_steps))) %>%
+  add_column(linear_effect_means = c(linear_effect_means, rep(NA, n_forecast_steps))) %>%
+  add_column(linear_effect_sds = c(linear_effect_sds, rep(NA, n_forecast_steps)))
+
+ggplot(components_df, aes(x = month, y = n)) +
+  geom_line(aes(y = seasonal_effect_means), color = "orange") +
+  geom_ribbon(
+    aes(
+      ymin = seasonal_effect_means - 2 * seasonal_effect_sds,
+      ymax = seasonal_effect_means + 2 * seasonal_effect_sds
+    ),
+    alpha = 0.2,
+    fill = "orange"
+  ) +
+  theme(axis.title = element_blank()) +
+  geom_line(aes(y = linear_effect_means), color = "green") +
+  geom_ribbon(
+    aes(
+      ymin = linear_effect_means - 2 * linear_effect_sds,
+      ymax = linear_effect_means + 2 * linear_effect_sds
+    ),
+    alpha = 0.2,
+    fill = "green"
+  ) +
+  theme(axis.title = element_blank())
+
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "AirPassengers, decomposition into a linear trend and a seasonal component (both centered)."}
+knitr::include_graphics("images/airp_components.png")
+```
+
+
+
+## Wrapping up
+
+We've seen how with DLMs, there is a bunch of interesting stuff you can do -- apart from obtaining forecasts, which probably will be the ultimate goal in most applications -- : You can inspect the smoothed and the filtered estimates from the Kálmán filter, and you can decompose a model into its posterior components. A particularly attractive model is _dynamic linear regression_, featured in our first example: It allows us to obtain regression coefficients that vary over time.
```

Comment:
```suggestion
We've seen how with DLMs, there is a bunch of interesting stuff you can do -- apart from obtaining forecasts, which probably will be the ultimate goal in most applications -- : You can inspect the smoothed and the filtered estimates from the Kálmán filter, and you can decompose a model into its posterior components. A particularly attractive model is _dynamic linear regression_, featured in our first example, which allows us to obtain regression coefficients that vary over time.
```

Reply:
The colon-initial-caps-sentence construct is pretty rare in English writing, so I killed off its second use.